pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -239,6 +241,32 @@ func (m *ManagerImpl) Start(activePods ActivePodsFunc, sourcesReady config.Sourc
 	return nil
 }
 
+// GetWatcherCallback returns callback function to be registered with plugin watcher
+func (m *ManagerImpl) GetWatcherCallback() watcher.RegisterCallbackFn {
+	return func(name string, endpoint string, versions []string, sockPath string) (error, chan bool) {
+		chanForAckOfNotification := make(chan bool)
+		if !m.isVersionCompatibleWithPlugin(versions) {
+			return fmt.Errorf("manager version, %s, is not among plugin supported versions %v", pluginapi.Version, versions), chanForAckOfNotification
+		}
+
+		if !v1helper.IsExtendedResourceName(v1.ResourceName(name)) {
+			errorString := fmt.Sprintf(errInvalidResourceName, name)
+			return fmt.Errorf("invalid name of device plugin socket: %v", errorString), chanForAckOfNotification
+		}
+
+		return m.addEndpointProbeMode(name, sockPath)
```

Comment:
I am trying to think what is the best-practice upgrade story for an existing device plugin. Given that a device plugin daemonset can run in a cluster with both old version and new version kubelet nodes, it seems a device plugin may first create a socket under /var/lib/kubelet/plugins, wait a bit to see whether kubelet pluginwatcher may contact it, and if not after a bound time, removes that socket and falls back to create a socket under /var/lib/kubelet/device_plugins/ and issue a Register grpc call. This unfortunately is both hacky and racy :(. That also means that on kubelet side, we need to carefully handle the case that addEndpointProbeMode and addEndpoint may be called in interleaving orders, and we need to make sure we only succeed one but fail the other.

I wonder whether we may provide some simple signal from kubelet side instead, e.g., creating a DEPRECATION file under /var/lib/kubelet/device_plugins/ directory. A device plugin on a node can then check the existence of this file to decide whether it should use probe mode or register mode. We can add this to the 1.12 release note and tell people that a device plugin on a node should use either mode but not both. What do you think?

Reply:
I like the later one i.e DEPRECATION file and mentioning in release notes. Agree that first one seems both racy and hacky and we may end up wasting energy in stabilising the code which is to be removed eventually.